File: Labs/Lab8/Problem.java
package Labs.Lab8;

public class Problem {
    public int coinChange(int[] coins, int amount) {
        
        int max = amount + 1;
        
        int[] dp = new int[amount + 1];
        
        for (int i = 0; i <= amount; i++) {
            dp[i] = max;
        }
        
        dp[0] = 0;

        
        for (int i = 1; i <= amount; i++) {
            for (int coin : coins) {
                if (coin <= i) {
                    dp[i] = Math.min(dp[i], dp[i - coin] + 1);
                }
            }
        }

        return dp[amount] > amount ? -1 : dp[amount];
    }
    
}

File: Labs/Lab8/problem.py
def coinChange(self, coins: List[int], amount: int) -> int:
    max_amount = amount + 1
    dp = [max_amount] * (amount + 1)
    dp[0] = 0

    for i in range(1, amount + 1):
        for coin in coins:
            if coin <= i:
                dp[i] = min(dp[i], dp[i - coin] + 1)
    return dp[amount] if dp[amount] != max_amount else -1
